File: src/main/java/no/sanchezrolfsen/framework/selenium/elements/Label.java
package no.sanchezrolfsen.framework.selenium.elements;

import no.sanchezrolfsen.framework.selenium.Browser;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import static java.lang.String.format;
import static no.sanchezrolfsen.framework.selenium.SeleniumUtils.safeIsVisible;
import static no.sanchezrolfsen.framework.selenium.SeleniumUtils.waitFor;

public class Label {
    final By labelSelector;

    public Label(WebElement element) {
        labelSelector = By.cssSelector(format("Label[for='%s']", element.getAttribute("id")));
    }

    public Label(String id) {
        labelSelector = By.cssSelector(format("Label[for='%s']", id));
    }

    public void click() {
        getDriver().findElement(labelSelector).click();
    }

    public void waitAndClick() {
        waitFor(labelSelector).click();
    }

    public String getText() {
        return getDriver().findElement(labelSelector).getText();
    }

    public boolean isDisplayed() {
        return safeIsVisible(labelSelector);
    }

    public WebElement getElement() {
        return getDriver().findElement(labelSelector);
    }

    private WebDriver getDriver() {
        return Browser.vanillaDriver();
    }
}

File: src/main/kotlin/no/sanchezrolfsen/framework/selenium/elements/Label.kt
package no.sanchezrolfsen.framework.selenium.elements

import no.sanchezrolfsen.framework.selenium.Browser
import no.sanchezrolfsen.framework.selenium.SeleniumUtils.safeIsVisible
import no.sanchezrolfsen.framework.selenium.SeleniumUtils.waitFor
import org.openqa.selenium.By
import org.openqa.selenium.WebDriver
import org.openqa.selenium.WebElement

class Label {
    val labelSelector: By

    constructor(element: WebElement) {
        labelSelector = By.cssSelector("Label[for='${element.getAttribute("id")}']")
    }

    constructor(id: String) {
        labelSelector = By.cssSelector("Label[for='$id']")
    }

    fun click() {
        getDriver().findElement(labelSelector).click()
    }

    fun waitAndClick() {
        waitFor(labelSelector).click()
    }

    fun getText(): String = getDriver().findElement(labelSelector).text

    fun isDisplayed(): Boolean = safeIsVisible(labelSelector)

    fun getElement(): WebElement = getDriver().findElement(labelSelector)

    private fun getDriver(): WebDriver = Browser.vanillaDriver()
}
